lib/services/role.go

```diff
@@ -118,78 +140,517 @@ type Access interface {
+		r.Spec.Allow.Logins = lcopy
+	} else {
+		r.Spec.Deny.Logins = lcopy
+	}
+}
+
+// GetNamespaces gets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) GetNamespaces(rct RoleConditionType) []string {
+	if rct == Allow {
+		return r.Spec.Allow.Namespaces
+	}
+	return r.Spec.Deny.Namespaces
+}
+
+// GetNamespaces sets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) SetNamespaces(rct RoleConditionType, namespaces []string) {
+	ncopy := utils.CopyStrings(namespaces)
+
+	if rct == Allow {
+		r.Spec.Allow.Namespaces = ncopy
+	} else {
+		r.Spec.Deny.Namespaces = ncopy
+	}
+}
+
+// GetNodeLabels gets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) GetNodeLabels(rct RoleConditionType) map[string]string {
+	if rct == Allow {
+		return r.Spec.Allow.NodeLabels
+	}
+	return r.Spec.Deny.NodeLabels
+}
+
+// SetNodeLabels sets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) SetNodeLabels(rct RoleConditionType, labels map[string]string) {
+	lcopy := utils.CopyStringMap(labels)
+
+	if rct == Allow {
+		r.Spec.Allow.NodeLabels = lcopy
+	} else {
+		r.Spec.Deny.NodeLabels = lcopy
+	}
+}
+
+// GetRules gets all allow or deny rules.
+func (r *RoleV3) GetRules(rct RoleConditionType) map[string][]string {
+	if rct == Allow {
+		return r.Spec.Allow.Rules
+	}
+	return r.Spec.Deny.Rules
+}
+
+// SetRules sets an allow or deny rule.
+func (r *RoleV3) SetRules(rct RoleConditionType, rrs map[string][]string) {
+	rcopy := utils.CopyStringMapSlices(rrs)
+
+	if rct == Allow {
+		r.Spec.Allow.Rules = rcopy
+	} else {
+		r.Spec.Deny.Rules = rcopy
+	}
+}
+
+// GetSystemResources gets all low level resources.
+func (r *RoleV3) GetSystemResources(rct RoleConditionType) map[string][]string {
+	if rct == Allow {
+		return r.Spec.Allow.SystemResources
+	}
+	return r.Spec.Deny.SystemResources
+}
+
+// SetSystemResources sets all low level resources.
+func (r *RoleV3) SetSystemResources(rct RoleConditionType, sr map[string][]string) {
+	srcopy := utils.CopyStringMapSlices(sr)
+
+	if rct == Allow {
+		r.Spec.Allow.SystemResources = srcopy
+	} else {
+		r.Spec.Deny.SystemResources = srcopy
+	}
+}
+
+// Check checks validity of all parameters and sets defaults
+func (r *RoleV3) CheckAndSetDefaults() error {
+	// make sure we have defaults for all fields
+	if r.Metadata.Name == "" {
+		return trace.BadParameter("missing parameter Name")
+	}
+	if r.Metadata.Namespace == "" {
+		r.Metadata.Namespace = defaults.Namespace
+	}
+	if r.Spec.Options == nil {
+		r.Spec.Options = map[string]interface{}{
+			MaxSessionTTL: NewDuration(defaults.MaxCertDuration),
+		}
+	}
+	if r.Spec.Allow.Namespaces == nil {
+		r.Spec.Allow.Namespaces = []string{Wildcard}
+	}
+	if r.Spec.Allow.NodeLabels == nil {
+		r.Spec.Allow.NodeLabels = map[string]string{Wildcard: Wildcard}
+	}
+	if r.Spec.Allow.SystemResources == nil {
+		r.Spec.Allow.SystemResources = DefaultUserSystemResources
```

Comment:
why does this apply user defaults here and when we create a user object as well? seems redundant.